File: CAVtron/arduino/src/serial_comm.cpp
/**
 * @file serial_comm.cpp
 * @brief Implementação do protocolo de comunicação serial com CRC-8
 */

#include "serial_comm.h"
#include <string.h>
#include <stdio.h>

// ============================================================
//  INICIALIZAÇÃO
// ============================================================

void SerialComm::begin() {
    SERIAL_RPI.begin(BAUD_RATE_RPI);
    SERIAL_DEBUG.begin(BAUD_RATE_DEBUG);

    _parseState       = PARSE_WAIT_START;
    _rxIndex          = 0;
    _lastRxTimestamp  = millis();
    _lastMsgTimestamp = millis();

    memset(&_lastMsg, 0, sizeof(_lastMsg));

    SERIAL_DEBUG.println(F("[SerialComm] Inicializada."));
}

// ============================================================
//  UPDATE — PARSER INCREMENTAL (FSM)
// ============================================================

bool SerialComm::update() {
    bool newMessageReady = false;

    while (SERIAL_RPI.available() > 0) {
        const char c = (char)SERIAL_RPI.read();

        switch (_parseState) {
            // ---- Aguarda delimitador de início '<' ----
            case PARSE_WAIT_START:
                if (c == MSG_START_CHAR) {
                    _rxIndex    = 0;
                    _parseState = PARSE_READING;
                }
                break;

            // ---- Lê conteúdo até '>' ----
            case PARSE_READING:
                if (c == MSG_END_CHAR) {
                    // Fim do frame: null-termina e processa
                    _rxBuffer[_rxIndex] = '\0';
                    _parseState = PARSE_WAIT_START;

                    if (_parseBuffer()) {
                        _lastRxTimestamp  = millis();
                        _lastMsgTimestamp = millis();
                        newMessageReady   = true;
                    }
                } else if (_rxIndex < (MSG_MAX_LENGTH - 1)) {
                    _rxBuffer[_rxIndex++] = c;
                } else {
                    // Buffer overflow: descarta frame corrompido
                    SERIAL_DEBUG.println(F("[SerialComm] WARN: Buffer overflow, frame descartado."));
                    _parseState = PARSE_WAIT_START;
                    _rxIndex    = 0;
                }
                break;

            default:
                _parseState = PARSE_WAIT_START;
                break;
        }
    }

    return newMessageReady;
}

// ============================================================
//  PARSER DO BUFFER
// ============================================================

bool SerialComm::_parseBuffer() {
    /**
     * Buffer contém: TIPO,PARAM1,...,PARAMN,CRCXX
     * Onde CRCXX é o CRC-8 em hexadecimal (2 dígitos uppercase)
     *
     * Estratégia:
     * 1. Encontrar o último ',' (separador antes do CRC)
     * 2. Extrair e converter o CRC hexadecimal
     * 3. Calcular CRC sobre o conteúdo antes do último ','
     * 4. Tokenizar os campos
     */
    memset(&_lastMsg, 0, sizeof(_lastMsg));

    // Encontra a última vírgula (antes do campo CRC)
    char* lastComma = strrchr(_rxBuffer, MSG_DELIMITER);
    if (!lastComma) {
        SERIAL_DEBUG.println(F("[SerialComm] ERRO: Formato de frame inválido."));
        return false;
    }

    // Extrai CRC recebido (hexadecimal)
    const char* crcStr = lastComma + 1;
    _lastMsg.receivedCRC = (uint8_t)strtol(crcStr, nullptr, 16);

    // Calcula CRC sobre o conteúdo antes do último ','
    const uint8_t contentLen = (uint8_t)(lastComma - _rxBuffer);
    _lastMsg.calculatedCRC   = _crc8((const uint8_t*)_rxBuffer, contentLen);

    // Verifica integridade
    if (_lastMsg.receivedCRC != _lastMsg.calculatedCRC) {
        SERIAL_DEBUG.print(F("[SerialComm] ERRO CRC: recebido=0x"));
        SERIAL_DEBUG.print(_lastMsg.receivedCRC, HEX);
        SERIAL_DEBUG.print(F(" calculado=0x"));
        SERIAL_DEBUG.println(_lastMsg.calculatedCRC, HEX);
        _lastMsg.valid = false;
        return false;
    }

    _lastMsg.valid = true;

    // Tokeniza o conteúdo (sem o campo CRC)
    char content[MSG_MAX_LENGTH];
    strncpy(content, _rxBuffer, contentLen);
    content[contentLen] = '\0';

    // Extrai tipo (primeiro token)
    char* token = strtok(content, ",");
    if (!token) return false;
    strncpy(_lastMsg.type, token, sizeof(_lastMsg.type) - 1);

    // Extrai parâmetros restantes
    _lastMsg.paramCount = 0;
    while ((token = strtok(nullptr, ",")) != nullptr
           && _lastMsg.paramCount < 8)
    {
        strncpy(_lastMsg.params[_lastMsg.paramCount],
                token,
                sizeof(_lastMsg.params[0]) - 1);
        _lastMsg.paramCount++;
    }

    return true;
}

// ============================================================
//  ENVIO DE MENSAGENS
// ============================================================

void SerialComm::sendSensorData(float linePos,
                                bool  lineDetected,
                                bool  victimDetected,
                                float victimX,
                                float victimY)
{
    char content[MSG_MAX_LENGTH];
    snprintf(content, sizeof(content),
             "SENSOR,%.3f,%d,%d,%.3f,%.3f",
             linePos,
             (int)lineDetected,
             (int)victimDetected,
             victimX,
             victimY);
    _sendFrame(content);
}

void SerialComm::sendStateChange(RobotState_t newState) {
    char content[32];
    snprintf(content, sizeof(content), "STATE_CHG,%d", (int)newState);
    _sendFrame(content);
}

void SerialComm::sendAck(const char* cmdType) {
    char content[32];
    snprintf(content, sizeof(content), "ACK,%s", cmdType);
    _sendFrame(content);
}

void SerialComm::sendPong() {
    _sendFrame("PONG");
}

void SerialComm::sendError(ErrorCode_t errorCode) {
    char content[32];
    snprintf(content, sizeof(content), "ERROR,%d", (int)errorCode);
    _sendFrame(content);
}

// ============================================================
//  WATCHDOG
// ============================================================

bool SerialComm::isWatchdogExpired() const {
    return (millis() - _lastMsgTimestamp) > WATCHDOG_TIMEOUT_MS;
}

void SerialComm::resetWatchdog() {
    _lastMsgTimestamp = millis();
}

// ============================================================
//  MÉTODOS PRIVADOS
// ============================================================

void SerialComm::_sendFrame(const char* content) {
    const uint8_t crc = _crc8str(content);

    // Formato: <CONTENT,CRCXX>\n
    SERIAL_RPI.print(MSG_START_CHAR);
    SERIAL_RPI.print(content);
    SERIAL_RPI.print(MSG_DELIMITER);

    // CRC em hexadecimal com padding de 2 dígitos
    if (crc < 0x10) SERIAL_RPI.print('0');
    SERIAL_RPI.print(crc, HEX);

    SERIAL_RPI.print(MSG_END_CHAR);
    SERIAL_RPI.print('\n');
}

uint8_t SerialComm::_crc8(const uint8_t* data, uint8_t len) {
    /**
     * CRC-8, polinômio gerador: x^8 + x^2 + x + 1 → 0x07
     * Padrão Dallas/Maxim — amplamente utilizado em sistemas embarcados.
     *
     * Algoritmo bit-a-bit (sem lookup table para economizar RAM):
     *   Para cada byte: XOR com registrador, depois 8 iterações de
     *   deslocamento com XOR condicional pelo polinômio.
     */
    uint8_t crc = 0x00;
    for (uint8_t i = 0; i < len; i++) {
        crc ^= data[i];
        for (uint8_t bit = 0; bit < 8; bit++) {
            if (crc & 0x80) {
                crc = (crc << 1) ^ 0x07;
            } else {
                crc <<= 1;
            }
        }
    }
    return crc;
}

uint8_t SerialComm::_crc8str(const char* str) {
    return _crc8((const uint8_t*)str, (uint8_t)strlen(str));
}

File: CAVtron/raspberry/serial_comm.py
"""
communication/serial_comm.py
============================
Protocolo de comunicação serial bidirecional com o Arduino.

Implementa o mesmo protocolo framed com CRC-8 definido na camada Arduino,
garantindo interoperabilidade direta sem camada de adaptação.

Formato do frame:
    <TIPO,PARAM1,...,PARAMN,HEXCRC>

Arquitetura de threading:
    Thread de leitura dedicada (daemon) lê bytes continuamente e
    alimenta uma fila thread-safe (queue.Queue). A thread principal
    consome a fila e despacha callbacks registrados, garantindo que
    a recepção serial não seja bloqueada por processamento de alto nível.

Referência CRC-8:
    Polinômio 0x07 (Dallas/Maxim) — idêntico ao implementado no Arduino.
"""

import serial
import threading
import queue
import time
import logging
from dataclasses import dataclass, field
from typing import Callable, Dict, List, Optional

logger = logging.getLogger(__name__)


# ============================================================
#  ESTRUTURAS DE DADOS
# ============================================================

@dataclass
class ParsedMessage:
    """Mensagem decodificada e validada pelo parser serial."""
    msg_type: str = ''
    params: List[str] = field(default_factory=list)
    received_crc: int = 0
    calculated_crc: int = 0
    valid: bool = False
    timestamp: float = field(default_factory=time.monotonic)


@dataclass
class SensorData:
    """Dados de sensores recebidos do Arduino via protocolo SENSOR."""
    line_position: float = 0.0      # [-1.0, +1.0]
    line_detected: bool = False
    victim_detected: bool = False
    victim_x: float = 0.0           # [-1.0, +1.0] posição normalizada
    victim_y: float = 0.0           # [-1.0, +1.0] posição normalizada
    timestamp: float = field(default_factory=time.monotonic)


# ============================================================
#  CRC-8 (polinômio 0x07, Dallas/Maxim)
# ============================================================

def crc8(data: bytes) -> int:
    """
    Calcula CRC-8 com polinômio gerador x^8 + x^2 + x + 1 (0x07).

    Algoritmo bit-a-bit idêntico ao firmware Arduino. Sem lookup table
    para reduzir footprint de memória e garantir portabilidade.

    Args:
        data: Bytes sobre os quais calcular o CRC.
    Returns:
        Byte de CRC calculado [0-255].
    """
    crc = 0x00
    for byte in data:
        crc ^= byte
        for _ in range(8):
            crc = ((crc << 1) ^ 0x07) & 0xFF if (crc & 0x80) else (crc << 1) & 0xFF
    return crc


def crc8_str(s: str) -> int:
    """Wrapper de crc8 para strings ASCII."""
    return crc8(s.encode('ascii'))


# ============================================================
#  CLASSE PRINCIPAL
# ============================================================

class SerialComm:
    """
    Gerenciador do protocolo serial Raspberry Pi <-> Arduino.

    Args:
        port       : Porta serial (ex: '/dev/serial0', '/dev/ttyUSB0')
        baudrate   : Taxa de transmissão (deve coincidir com Arduino)
        timeout_s  : Timeout de leitura serial em segundos
        watchdog_s : Intervalo máximo sem mensagem antes de alarme
    """

    MSG_START = ord('<')
    MSG_END   = ord('>')
    MSG_SEP   = ','

    def __init__(self,
                 port: str = '/dev/serial0',
                 baudrate: int = 115200,
                 timeout_s: float = 1.0,
                 watchdog_s: float = 2.0):
        self._port       = port
        self._baudrate   = baudrate
        self._timeout_s  = timeout_s
        self._watchdog_s = watchdog_s

        self._serial: Optional[serial.Serial] = None
        self._rx_queue: "queue.Queue[ParsedMessage]" = queue.Queue(maxsize=128)
        self._callbacks: Dict[str, List[Callable]] = {}
        self._rx_buffer = bytearray()
        self._reading   = False

        self._last_rx_time: float = time.monotonic()
        self._connected: bool     = False
        self._tx_lock = threading.Lock()

        self._rx_thread = threading.Thread(
            target=self._rx_loop,
            name='SerialRxThread',
            daemon=True
        )

    # ----------------------------------------------------------
    #  CICLO DE VIDA
    # ----------------------------------------------------------

    def connect(self) -> bool:
        """
        Abre a porta serial e inicia a thread de recepção.

        Returns:
            True se conexão estabelecida com sucesso.
        """
        try:
            self._serial = serial.Serial(
                port=self._port,
                baudrate=self._baudrate,
                timeout=self._timeout_s,
                bytesize=serial.EIGHTBITS,
                parity=serial.PARITY_NONE,
                stopbits=serial.STOPBITS_ONE
            )
            time.sleep(0.1)   # Aguarda estabilização do UART
            self._serial.reset_input_buffer()
            self._connected    = True
            self._last_rx_time = time.monotonic()
            self._rx_thread.start()
            logger.info(f"[SerialComm] Conectado em {self._port} @ {self._baudrate} baud")
            return True
        except serial.SerialException as e:
            logger.error(f"[SerialComm] Falha ao abrir porta: {e}")
            return False

    def disconnect(self):
        """Encerra a comunicação serial de forma segura."""
        self._connected = False
        if self._serial and self._serial.is_open:
            self._serial.close()
        logger.info("[SerialComm] Desconectado.")

    # ----------------------------------------------------------
    #  REGISTRO DE CALLBACKS
    # ----------------------------------------------------------

    def register_callback(self, msg_type: str,
                          callback: Callable[["ParsedMessage"], None]):
        """
        Registra callback para um tipo de mensagem.

        O callback é invocado na thread principal via process_queue().
        Múltiplos callbacks podem ser registrados para o mesmo tipo.

        Args:
            msg_type : Tipo da mensagem em maiúsculas (ex: 'SENSOR')
            callback : Função que recebe ParsedMessage como argumento
        """
        self._callbacks.setdefault(msg_type, []).append(callback)
        logger.debug(f"[SerialComm] Callback registrado: '{msg_type}'")

    # ----------------------------------------------------------
    #  PROCESSAMENTO DA FILA (LOOP PRINCIPAL)
    # ----------------------------------------------------------

    def process_queue(self, max_messages: int = 20):
        """
        Consome mensagens da fila e despacha callbacks.

        Deve ser chamado periodicamente no loop principal.
        Limita max_messages por chamada para não monopolizar CPU.

        Args:
            max_messages: Limite de mensagens processadas por invocação.
        """
        for _ in range(max_messages):
            try:
                msg = self._rx_queue.get_nowait()
                self._dispatch(msg)
            except queue.Empty:
                break

    def _dispatch(self, msg: ParsedMessage):
        for cb in self._callbacks.get(msg.msg_type, []):
            try:
                cb(msg)
            except Exception as e:
                logger.error(f"[SerialComm] Erro no callback '{msg.msg_type}': {e}")

    # ----------------------------------------------------------
    #  ENVIO DE COMANDOS
    # ----------------------------------------------------------

    def send_move(self, direction: str, speed: int):
        """
        Envia comando de movimento ao Arduino.

        Args:
            direction : 'FWD' | 'BWD' | 'ROT_L' | 'ROT_R' | 'STOP'
            speed     : Velocidade PWM [0-255]
        """
        self._send_frame(f"MOVE,{direction},{int(speed)}")

    def send_stop(self):
        """Para todos os motores imediatamente."""
        self._send_frame("STOP")

    def send_grip(self, action: str):
        """
        Envia comando de acionamento da garra.

        Args:
            action : 'OPEN' | 'CLOSE' | 'CAPTURE'
        """
        self._send_frame(f"GRIP,{action}")

    def send_lift(self, action: str):
        """
        Envia comando de elevação do braço.

        Args:
            action : 'UP' | 'DOWN'
        """
        self._send_frame(f"LIFT,{action}")

    def send_state(self, state_id: int):
        """Comanda adoção de novo estado FSM no Arduino."""
        self._send_frame(f"STATE,{state_id}")

    def send_ping(self):
        """Envia PING para verificar conectividade e resetar watchdog."""
        self._send_frame("PING")

    # ----------------------------------------------------------
    #  DIAGNÓSTICO
    # ----------------------------------------------------------

    def is_watchdog_expired(self) -> bool:
        """Retorna True se não houve RX válido por mais de watchdog_s."""
        return (time.monotonic() - self._last_rx_time) > self._watchdog_s

    def is_connected(self) -> bool:
        return self._connected and self._serial is not None and self._serial.is_open

    # ----------------------------------------------------------
    #  THREAD DE RECEPÇÃO (DAEMON)
    # ----------------------------------------------------------

    def _rx_loop(self):
        """
        Loop de leitura serial em thread daemon.

        FSM de parsing incremental:
          Aguarda '<' -> acumula bytes -> ao receber '>' processa frame.
        """
        logger.debug("[SerialRxThread] Iniciada.")
        while self._connected:
            try:
                if not (self._serial and self._serial.is_open):
                    time.sleep(0.05)
                    continue

                raw_byte = self._serial.read(1)
                if not raw_byte:
                    continue

                b = raw_byte[0]

                if b == self.MSG_START:
                    self._rx_buffer.clear()
                    self._reading = True
                elif b == self.MSG_END and self._reading:
                    self._reading = False
                    raw = self._rx_buffer.decode('ascii', errors='ignore')
                    msg = self._parse(raw)
                    if msg.valid:
                        self._last_rx_time = time.monotonic()
                        try:
                            self._rx_queue.put_nowait(msg)
                        except queue.Full:
                            logger.warning("[SerialComm] Fila RX cheia.")
                elif self._reading:
                    if len(self._rx_buffer) < 128:
                        self._rx_buffer.append(b)
                    else:
                        self._rx_buffer.clear()
                        self._reading = False
                        logger.warning("[SerialComm] Buffer overflow.")

            except serial.SerialException as e:
                logger.error(f"[SerialRxThread] Erro serial: {e}")
                time.sleep(0.1)
            except Exception as e:
                logger.error(f"[SerialRxThread] Exceção: {e}")
                time.sleep(0.1)

    # ----------------------------------------------------------
    #  PARSER
    # ----------------------------------------------------------

    def _parse(self, raw: str) -> ParsedMessage:
        """
        Faz o parse do conteúdo de um frame recebido.

        1. Encontra o último ',' (separador antes do campo CRC)
        2. Extrai e converte o CRC hexadecimal
        3. Valida CRC sobre o conteúdo anterior
        4. Tokeniza tipo e parâmetros

        Args:
            raw: String entre '<' e '>' sem os delimitadores.
        Returns:
            ParsedMessage com valid=True se CRC confere.
        """
        msg = ParsedMessage()
        last_comma = raw.rfind(self.MSG_SEP)
        if last_comma == -1:
            return msg

        content = raw[:last_comma]
        crc_str = raw[last_comma + 1:].strip()

        try:
            msg.received_crc = int(crc_str, 16)
        except ValueError:
            return msg

        msg.calculated_crc = crc8_str(content)
        if msg.received_crc != msg.calculated_crc:
            logger.warning(
                f"[SerialComm] CRC mismatch | "
                f"rx=0x{msg.received_crc:02X} calc=0x{msg.calculated_crc:02X}"
            )
            return msg

        tokens = content.split(self.MSG_SEP)
        if not tokens:
            return msg

        msg.msg_type = tokens[0].strip().upper()
        msg.params   = [t.strip() for t in tokens[1:]]
        msg.valid    = True
        return msg

    # ----------------------------------------------------------
    #  ENVIO INTERNO
    # ----------------------------------------------------------

    def _send_frame(self, content: str):
        """
        Formata e envia frame com CRC ao Arduino.

        Formato: <CONTENT,HEXCRC>\n
        """
        if not self.is_connected():
            logger.warning("[SerialComm] Envio sem conexão ativa.")
            return

        crc   = crc8_str(content)
        frame = f"<{content},{crc:02X}>\n".encode('ascii')
        with self._tx_lock:
            try:
                self._serial.write(frame)
                logger.debug(f"[SerialComm] TX: {frame.decode().strip()}")
            except serial.SerialException as e:
                logger.error(f"[SerialComm] Erro ao enviar: {e}")
